fungible_proxy/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+    use std::convert::TryInto;
+
+    fn alice_account() -> AccountId {
+        "alice.near".to_string()
+    }
+
+    fn get_context(
+        predecessor_account_id: AccountId,
+        attached_deposit: Balance,
+        prepaid_gas: Gas,
+        is_view: bool,
+    ) -> VMContext {
+        VMContext {
+            current_account_id: predecessor_account_id.clone(),
+            signer_account_id: predecessor_account_id.clone(),
+            signer_account_pk: vec![0, 1, 2],
+            predecessor_account_id,
+            input: vec![],
+            block_index: 1,
+            block_timestamp: 0,
+            epoch_height: 1,
+            account_balance: 0,
+            account_locked_balance: 0,
+            storage_usage: 10u64.pow(6),
+            attached_deposit,
+            prepaid_gas,
+            random_seed: vec![0, 1, 2],
+            is_view,
+            output_data_receivers: vec![],
+        }
+    }
+
+    fn ntoy(near_amount: Balance) -> Balance {
+        near_amount * 10u128.pow(24)
+    }
+
+    /// Helper function: get default values for PaymentArgs
+    fn get_default_payment_args() -> PaymentArgs {
+        PaymentArgs {
+            fee_address: "fee.requestfinance.near".to_string().try_into().unwrap(),
+            fee_amount: 200.into(),
+            payment_reference: "abc7c8bb1234fd12".into(),
+            to: "dummy.payee.near".to_string().try_into().unwrap(),
+        }
+    }
+
+    /// Helper function: convert a PaymentArgs into the msg string to be passed into `ft_transfer_call`
+    fn get_msg_from_args(args: PaymentArgs) -> String {
+        serde_json::to_string(&args).unwrap()
+    }
+
+    #[test]
+    #[should_panic(expected = r#"Incorrect payment reference length"#)]
+    fn transfer_with_invalid_reference_length() {
+        let context = get_context(alice_account(), ntoy(100), MIN_GAS, false);
+        testing_env!(context);
+        let mut contract = FungibleProxy::default();
+
+        let mut args = get_default_payment_args();
+        args.payment_reference = "0x11223344556677".to_string();
+        let msg = get_msg_from_args(args);
+
+        contract.ft_on_transfer(alice_account(), "1".into(), msg);
+    }
+
+    #[test]
+    #[should_panic(expected = r#"Payment reference value error"#)]
+    fn transfer_with_invalid_reference_value() {
+        let context = get_context(alice_account(), ntoy(100), MIN_GAS, false);
+        testing_env!(context);
+        let mut contract = FungibleProxy::default();
+
+        let mut args = get_default_payment_args();
+        args.payment_reference = "0x123".to_string();
+        let msg = get_msg_from_args(args);
+
+        contract.ft_on_transfer(alice_account(), "1".into(), msg);
+    }
+
+    #[test]
+    #[should_panic(expected = r#"Not enough attached Gas to call this method"#)]
+    fn transfer_with_not_enough_gas() {
+        let context = get_context(alice_account(), ntoy(100), 10u64.pow(14), false);
```

Comment:
Consider, for clarity
```suggestion
        let context = get_context(alice_account(), ntoy(100), MIN_GAS - 1, false);
```